packages/ariakit-utils/src/__tests__/dom-test.ts

```diff
@@ -0,0 +1,601 @@
+
+  expect(isTextField(div)).toEqual(false);
+  expect(isTextField(input)).toEqual(true);
+  expect(isTextField(input2)).toEqual(false);
+  expect(isTextField(textarea)).toEqual(true);
+
+  const selectionStartSpy = jest
+    .spyOn(input, "selectionStart", "get")
+    .mockImplementation(() => {
+      throw new Error();
+    });
+
+  expect(isTextField(input)).toEqual(false);
+
+  selectionStartSpy.mockRestore();
+
+  // Reset
+  document.body.innerHTML = initialInnerHTML;
+});
+
+test("getPopupRole", () => {
+  const initialInnerHTML = document.body.innerHTML;
+  document.body.innerHTML = `
+    <div id="testNode" role="dialog" />
+    <div id="testNode2" role="menu" />
+    <div id="testNode3" role="listbox" />
+    <div id="testNode4" role="tree" />
+    <div id="testNode5" role="grid" />
+    <div id="testNode6" role="unsupported" />
+    <div id="testNode7" />
+  `;
+  const div = document.getElementById("testNode") as HTMLElement;
+  const div2 = document.getElementById("testNode2") as HTMLElement;
+  const div3 = document.getElementById("testNode3") as HTMLElement;
+  const div4 = document.getElementById("testNode4") as HTMLElement;
+  const div5 = document.getElementById("testNode5") as HTMLElement;
+  const div6 = document.getElementById("testNode6") as HTMLElement;
+  const div7 = document.getElementById("testNode7") as HTMLElement;
+
+  expect(getPopupRole(div)).toEqual("dialog");
+  expect(getPopupRole(div2)).toEqual("menu");
+  expect(getPopupRole(div3)).toEqual("listbox");
+  expect(getPopupRole(div4)).toEqual("tree");
+  expect(getPopupRole(div5)).toEqual("grid");
+  expect(getPopupRole(div6)).toEqual(undefined);
+  expect(getPopupRole(div6, "dialog")).toEqual("dialog");
+  expect(getPopupRole(div6, true)).toEqual(true);
+  expect(getPopupRole(div7)).toEqual(undefined);
+  expect(getPopupRole(div7, "dialog")).toEqual("dialog");
+
+  // Reset
+  document.body.innerHTML = initialInnerHTML;
+});
+
+test("getTextboxSelection", () => {
+  const initialInnerHTML = document.body.innerHTML;
+  document.body.innerHTML = `
+    <input id="testInput" type="text" value="some text" />
+    <div contenteditable="true" id="testContentEditable" />
+  `;
+  const input = document.getElementById("testInput") as HTMLInputElement;
+  input.setSelectionRange(2, 5);
+  expect(getTextboxSelection(input)).toEqual({
+    start: 2,
+    end: 5,
+  });
+
+  // TODO: Figure out how to test the contenteditable case
+  // This https://github.com/jsdom/jsdom/issues/1670#issuecomment-267843544
+  //  implies that maybe content editiable isn't well supported in jsdom
+
+  // Reset
+  document.body.innerHTML = initialInnerHTML;
+});
+
+test("scrollIntoViewIfNeeded", () => {
+  // TODO: Figure out how to test this
+  // Seems like JSDOM doesn't support scroll stuff because it doesn't really do the layout
+  // https://github.com/jsdom/jsdom/issues/1695
+  // It seems like maybe we could fake a bunch of functions but it might not have a lot of value anyways
```

Comment:
Let's ignore what we can't test with JSDOM. We can test it in the future with some examples (on the ariakit package) using e2e tests (not set up yet).